File: VillagesTown/Engines/TurnEngine.swift
//
//  TurnEngine.swift
//  VillagesTown
//
//  Created by Furkan Kaynar on 12.04.2020.
//  Copyright © 2020 Furkan Kaynar. All rights reserved.
//

import Foundation

class TurnEngine {
    let populationEngine = PopulationEngine()
    let unitUpkeepEngine = UnitUpkeepEngine()
    let movementEngine = MovementEngine()
    let aiEngine = AIEngine()

    func doTurn() {
        let game = GameManager.shared
        game.currentTurn += 1
        game.clearTurnEvents()

        print("\n" + String(repeating: "=", count: 60))
        print("🎲 TURN \(game.currentTurn)")
        print(String(repeating: "=", count: 60) + "\n")

        // 1. Building Production
        print("\n📦 Phase 1: Building Production")
        doBuildingProduction()

        // 2. Tax Collection
        print("\n💰 Phase 2: Tax Collection")
        collectTaxes()

        // 3. Army Upkeep (replaces unit upkeep)
        print("\n💸 Phase 3: Army Upkeep")
        processArmyUpkeep()

        // 4. Population & Happiness
        print("\n👥 Phase 4: Population & Happiness")
        processPopulation()

        // 5. Update happiness
        processHappiness()

        // 6. Garrison Regeneration
        print("\n🛡️ Phase 6: Garrison Regeneration")
        processGarrisonRegeneration()

        // 7. Mid-route Army Interception
        print("\n🗡️ Phase 7: Army Interception")
        processArmyInterception()

        // 8. Process Army Movement & Combat at destinations
        print("\n⚔️ Phase 8: Army Movement & Combat")
        processArmyMovement()

        // 9. AI turns
        print("\n🤖 Phase 9: AI Turns")
        processAITurns()

        // 10. Detect incoming enemies
        print("\n👁️ Phase 10: Intelligence")
        detectIncomingEnemies()

        // 11. Check victory conditions
        print("\n🏆 Phase 11: Victory Check")
        checkVictory()

        print("\n" + String(repeating: "=", count: 60))
        print("✅ Turn \(game.currentTurn) Complete")
        printGameStatus()
        print(String(repeating: "=", count: 60) + "\n")
    }

    private func doBuildingProduction() {
        GameManager.shared.map.villages = GameManager.shared.map.villages.map { village in
            var mutableVillage = village
            BuildingProductionEngine.consumeAndProduceAll(in: &mutableVillage)
            return mutableVillage
        }

        // Sync all village resources to global pool
        GameManager.shared.syncGlobalResources()
    }

    private func collectTaxes() {
        var villages = GameManager.shared.map.villages
        populationEngine.collectTaxes(for: &villages)
        GameManager.shared.map.villages = villages

        // Sync to global pool
        GameManager.shared.syncGlobalResources()
    }

    private func processPopulation() {
        var villages = GameManager.shared.map.villages
        populationEngine.processPopulationGrowth(for: &villages)
        GameManager.shared.map.villages = villages
    }

    private func processHappiness() {
        var villages = GameManager.shared.map.villages
        populationEngine.processHappiness(for: &villages)
        GameManager.shared.map.villages = villages
    }

    private func processArmyUpkeep() {
        let game = GameManager.shared

        // Calculate total upkeep for each player's armies
        for player in game.players {
            var totalUpkeep: [Resource: Int] = [:]

            for army in game.getArmiesFor(playerID: player.id) {
                for unit in army.units {
                    let stats = Unit.getStats(for: unit.unitType)
                    for (resource, amount) in stats.upkeep {
                        totalUpkeep[resource, default: 0] += amount
                    }
                }
            }

            // Deduct upkeep
            for (resource, amount) in totalUpkeep {
                game.modifyGlobalResource(playerID: player.id, resource: resource, amount: -amount)
            }

            if !totalUpkeep.isEmpty {
                print("   \(player.name) army upkeep: \(totalUpkeep)")
            }
        }
    }

    private func processArmyMovement() {
        let game = GameManager.shared
        var arrivedArmies: [(Army, Village)] = []

        // Advance all marching armies
        for i in game.armies.indices {
            if game.armies[i].isMarching {
                game.armies[i].advanceMarch()

                // Check for arrival
                if !game.armies[i].isMarching,
                   let destID = game.armies[i].stationedAt,
                   let destination = game.map.villages.first(where: { $0.id == destID }) {
                    arrivedArmies.append((game.armies[i], destination))
                }
            }
        }

        // Process arrivals (combat if enemy, merge if friendly)
        for (army, destination) in arrivedArmies {
            if army.owner != destination.owner {
                // COMBAT!
                resolveCombat(attacker: army, at: destination)
            } else {
                // Friendly arrival - merge armies
                game.mergeArmiesAt(villageID: destination.id, owner: army.owner)
                game.addTurnEvent(.armyArrived(armyName: army.name, destination: destination.name))
                print("   \(army.name) arrived at friendly \(destination.name)")
            }
        }
    }

    private func resolveCombat(attacker: Army, at village: Village) {
        let game = GameManager.shared
        let combatEngine = CombatEngine()

        // Get defending armies
        let defenderArmies = game.getArmiesAt(villageID: village.id).filter { $0.owner == village.owner }
        var defenderUnits: [Unit] = defenderArmies.flatMap { $0.units }
        var attackerUnits = attacker.units

        // Resolve combat
        let result = combatEngine.resolveCombat(
            attackers: &attackerUnits,
            defenders: &defenderUnits,
            location: village.coordinates,
            map: game.map,
            defendingVillage: village
        )

        // Update attacker army
        if var mutableArmy = game.armies.first(where: { $0.id == attacker.id }) {
            mutableArmy.units = attackerUnits.filter { $0.isAlive }
            mutableArmy.removeDeadUnits()

            if mutableArmy.units.isEmpty {
                game.removeArmy(mutableArmy.id)
            } else {
                game.updateArmy(mutableArmy)
            }
        }

        // Update defender armies
        for defArmy in defenderArmies {
            game.removeArmy(defArmy.id)
        }
        // Create new defender army with survivors
        let survivingDefenders = defenderUnits.filter { $0.isAlive }
        if !survivingDefenders.isEmpty {
            _ = game.createArmy(units: survivingDefenders, stationedAt: village.id, owner: village.owner)
        }

        // Damage garrison
        var mutableVillage = village
        let garrisonDamage = result.attackerWon ? mutableVillage.garrisonStrength : mutableVillage.garrisonStrength / 2
        mutableVillage.damageGarrison(amount: garrisonDamage)
        game.updateVillage(mutableVillage)

        // Handle village conquest - need garrison to be 0 as well
        let canConquer = result.attackerWon && survivingDefenders.isEmpty && mutableVillage.garrisonStrength == 0

        if canConquer {
            let oldOwner = mutableVillage.owner
            mutableVillage.owner = attacker.owner
            mutableVillage.population = Int(Double(mutableVillage.population) * 0.7)
            mutableVillage.happiness -= 30
            mutableVillage.garrisonStrength = 3 // New owner gets small garrison
            game.updateVillage(mutableVillage)

            // Station attacking army at village
            if var attackingArmy = game.armies.first(where: { $0.id == attacker.id }) {
                attackingArmy.station(at: village.id)
                game.updateArmy(attackingArmy)
            }

            if attacker.owner == "player" {
                game.addTurnEvent(.villageConquered(villageName: village.name))
            } else if oldOwner == "player" {
                game.addTurnEvent(.villageLost(villageName: village.name))
            }
            print("   🏆 \(attacker.name) conquered \(village.name)!")
        } else if result.attackerWon {
            if attacker.owner == "player" {
                game.addTurnEvent(.battleWon(location: village.name, casualties: result.attackerCasualties))
            } else if village.owner == "player" {
                game.addTurnEvent(.battleLost(location: village.name, casualties: result.defenderCasualties))
            }
            print("   ⚔️ \(attacker.name) won battle at \(village.name) but defenders remain")
        } else {
            if attacker.owner == "player" {
                game.addTurnEvent(.battleLost(location: village.name, casualties: result.attackerCasualties))
            } else if village.owner == "player" {
                game.addTurnEvent(.battleWon(location: village.name, casualties: result.defenderCasualties))
            }
            print("   ❌ \(attacker.name) defeated at \(village.name)")
        }
    }

    private func detectIncomingEnemies() {
        let game = GameManager.shared
        let playerVillages = game.getPlayerVillages(playerID: "player")

        for army in game.armies where army.owner != "player" && army.isMarching {
            if let destID = army.destination,
               playerVillages.contains(where: { $0.id == destID }),
               let destVillage = game.map.villages.first(where: { $0.id == destID }) {
                game.addTurnEvent(.enemyApproaching(
                    enemyName: army.name,
                    target: destVillage.name,
                    turns: army.turnsUntilArrival
                ))
            }
        }
    }

    private func processGarrisonRegeneration() {
        let game = GameManager.shared

        // Regenerate garrison for all owned villages (not neutral)
        for i in game.map.villages.indices {
            if game.map.villages[i].owner != "neutral" {
                let oldGarrison = game.map.villages[i].garrisonStrength
                game.map.villages[i].regenerateGarrison()
                let newGarrison = game.map.villages[i].garrisonStrength

                if newGarrison > oldGarrison {
                    print("   \(game.map.villages[i].name): Garrison \(oldGarrison) → \(newGarrison)")
                }
            }
        }
    }

    private func processArmyInterception() {
        let game = GameManager.shared

        // Find pairs of armies traveling in opposite directions on the same route
        let marchingArmies = game.armies.filter { $0.isMarching }

        var processedPairs: Set<String> = []
        var interceptedArmyIDs: Set<UUID> = []

        for army1 in marchingArmies {
            for army2 in marchingArmies {
                // Skip same army or same owner
                guard army1.id != army2.id,
                      army1.owner != army2.owner,
                      !interceptedArmyIDs.contains(army1.id),
                      !interceptedArmyIDs.contains(army2.id) else {
                    continue
                }

                // Check if they're on the same route (opposite directions)
                let pairKey = [army1.id.uuidString, army2.id.uuidString].sorted().joined(separator: "-")
                guard !processedPairs.contains(pairKey) else { continue }
                processedPairs.insert(pairKey)

                // Check if army1's destination is army2's origin and vice versa
                let army1ToArmy2 = army1.destination == army2.origin && army1.origin == army2.destination
                let samePath = army1ToArmy2

                if samePath {
                    // They're on the same path going opposite directions - INTERCEPT!
                    print("   🗡️ INTERCEPTION: \(army1.name) vs \(army2.name)")

                    resolveFieldBattle(army1ID: army1.id, army2ID: army2.id)
                    interceptedArmyIDs.insert(army1.id)
                    interceptedArmyIDs.insert(army2.id)

                    // Add event
                    if army1.owner == "player" || army2.owner == "player" {
                        let enemyArmy = army1.owner == "player" ? army2 : army1
                        let playerArmy = army1.owner == "player" ? army1 : army2
                        game.addTurnEvent(.battleWon(location: "en route", casualties: 0))
                    }
                }
            }
        }
    }

    private func resolveFieldBattle(army1ID: UUID, army2ID: UUID) {
        let game = GameManager.shared
        let combatEngine = CombatEngine()

        guard var army1 = game.armies.first(where: { $0.id == army1ID }),
              var army2 = game.armies.first(where: { $0.id == army2ID }) else {
            return
        }

        var units1 = army1.units
        var units2 = army2.units

        // Field battle - no village bonuses
        let result = combatEngine.resolveCombat(
            attackers: &units1,
            defenders: &units2,
            location: CGPoint(x: 10, y: 10), // Middle of map
            map: game.map,
            defendingVillage: nil
        )

        // Update armies
        army1.units = units1.filter { $0.isAlive }
        army2.units = units2.filter { $0.isAlive }

        if army1.units.isEmpty {
            game.removeArmy(army1.id)
            print("   💀 \(army1.name) destroyed in field battle")
        } else {
            game.updateArmy(army1)
        }

        if army2.units.isEmpty {
            game.removeArmy(army2.id)
            print("   💀 \(army2.name) destroyed in field battle")
        } else {
            game.updateArmy(army2)
        }
    }

    private func processAITurns() {
        let game = GameManager.shared

        // Get all AI players
        let aiPlayers = game.players.filter { !$0.isHuman && !$0.isEliminated }

        if aiPlayers.isEmpty {
            print("No active AI players")
            return
        }

        // Each AI takes their turn
        for aiPlayer in aiPlayers {
            aiEngine.executeAITurn(for: aiPlayer, map: &game.map)
        }
    }

    private func checkVictory() {
        let game = GameManager.shared

        // Check each player
        for i in 0..<game.players.count {
            let playerVillages = game.map.villages.filter { $0.owner == game.players[i].id }

            // Eliminate if no villages
            if playerVillages.isEmpty && !game.players[i].isEliminated {
                game.players[i].isEliminated = true
                print("💀 \(game.players[i].name) has been eliminated!")
            }
        }

        // Check for overall victory
        let activePlayers = game.players.filter { !$0.isEliminated }

        if activePlayers.count == 1 {
            let winner = activePlayers[0]
            print("🎉🎉🎉 \(winner.name) HAS WON THE GAME! 🎉🎉🎉")
            print("Victory by Domination!")
        }

        // Check for player defeat
        let player = game.players.first(where: { $0.isHuman })
        if player?.isEliminated == true {
            print("💔 You have been defeated!")
        }
    }

    private func printGameStatus() {
        let game = GameManager.shared

        print("📊 Game Status:")
        for player in game.players {
            if player.isEliminated {
                print("   ❌ \(player.name): ELIMINATED")
            } else {
                let villages = game.map.villages.filter { $0.owner == player.id }
                let armies = game.getArmiesFor(playerID: player.id)
                let totalUnits = armies.reduce(0) { $0 + $1.unitCount }
                let marchingArmies = armies.filter { $0.isMarching }.count
                print("   ✓ \(player.name): \(villages.count) villages, \(armies.count) armies (\(totalUnits) units, \(marchingArmies) marching)")
            }
        }
        print("📊 Total armies: \(game.armies.count)")
    }

    // MARK: - Legacy compatibility
    static func doBuildingProduction(game: inout GameManager) {
        for index in game.map.villages.indices {
            BuildingProductionEngine.consumeAndProduceAll(in: &game.map.villages[index])
        }
    }

    static func doTurn(game: inout GameManager) {
        GameManager.shared.turnEngine.doTurn()
    }
}

File: villages_town_flutter/lib/engines/turn_engine.dart
import 'dart:math';
import '../data/models/army.dart';
import '../data/models/resource.dart';
import '../data/models/turn_event.dart';
import '../data/models/village.dart';
import 'ai_engine.dart';
import 'building_production_engine.dart';
import 'combat_engine.dart';
import 'game_manager.dart';
import 'population_engine.dart';

class TurnEngine {
  final PopulationEngine _populationEngine = PopulationEngine();
  final AIEngine _aiEngine = AIEngine();
  final CombatEngine _combatEngine = CombatEngine();

  void doTurn() {
    final game = GameManager.shared;
    game.currentTurn++;
    game.clearTurnEvents();

    // 0. Reset mobilization counters
    for (var i = 0; i < game.map.villages.length; i++) {
      game.map.villages[i].recruitsThisTurn = 0;
    }

    // 1. Building Production
    _doBuildingProduction();

    // 2. Tax Collection
    _collectTaxes();

    // 3. Army Upkeep
    _processArmyUpkeep();

    // 4. Population & Happiness
    _processPopulation();
    _processHappiness();

    // 6. Garrison Regeneration
    _processGarrisonRegeneration();

    // 7. Mid-route Army Interception
    _processArmyInterception();

    // 8. Army Movement & Combat
    _processArmyMovement();

    // 9. AI Turns
    _processAITurns();

    // 10. Intelligence
    _detectIncomingEnemies();

    // 11. Victory Check
    _checkVictory();

    // Notify is called by GameManager methods
  }

  void _doBuildingProduction() {
    final game = GameManager.shared;
    for (var i = 0; i < game.map.villages.length; i++) {
      BuildingProductionEngine.consumeAndProduceAll(game.map.villages[i]);
    }
    game.syncGlobalResources();
  }

  void _collectTaxes() {
    final game = GameManager.shared;
    final villages = game.map.villages.toList();
    _populationEngine.collectTaxes(villages);
    game.syncGlobalResources();
  }

  void _processArmyUpkeep() {
    final game = GameManager.shared;

    for (final player in game.players) {
      final totalUpkeep = <Resource, int>{};

      for (final army in game.getArmiesFor(player.id)) {
        for (final unit in army.units) {
          final stats = unit.unitType.stats;
          for (final entry in stats.upkeep.entries) {
            totalUpkeep[entry.key] = (totalUpkeep[entry.key] ?? 0) + entry.value;
          }
        }
      }

      for (final entry in totalUpkeep.entries) {
        game.modifyGlobalResource(player.id, entry.key, -entry.value);
      }
    }
  }

  void _processPopulation() {
    final game = GameManager.shared;
    final villages = game.map.villages.toList();
    _populationEngine.processPopulationGrowth(villages);
    game.map.villages = villages;
  }

  void _processHappiness() {
    final game = GameManager.shared;
    final villages = game.map.villages.toList();
    _populationEngine.processHappiness(villages);
    game.map.villages = villages;
  }

  void _processGarrisonRegeneration() {
    final game = GameManager.shared;
    for (var i = 0; i < game.map.villages.length; i++) {
      if (game.map.villages[i].owner != 'neutral') {
        game.map.villages[i].regenerateGarrison();
      }
    }
  }

  void _processArmyInterception() {
    final game = GameManager.shared;
    final marchingArmies = game.armies.where((a) => a.isMarching).toList();

    final processedPairs = <String>{};
    final interceptedArmyIds = <String>{};

    for (final army1 in marchingArmies) {
      for (final army2 in marchingArmies) {
        if (army1.id == army2.id) continue;
        if (army1.owner == army2.owner) continue;
        if (interceptedArmyIds.contains(army1.id)) continue;
        if (interceptedArmyIds.contains(army2.id)) continue;

        final pairKey = [army1.id, army2.id]..sort();
        final key = pairKey.join('-');
        if (processedPairs.contains(key)) continue;
        processedPairs.add(key);

        // Check if armies intercept:
        // 1. Same path opposite directions (will pass each other)
        final oppositeDirection = army1.destination == army2.origin && army1.origin == army2.destination;

        // 2. Same destination, close to each other
        final sameDestination = army1.destination == army2.destination;
        final bothCloseToArrival = army1.turnsUntilArrival <= 1 && army2.turnsUntilArrival <= 1;

        // 3. One army's destination is the other's origin (pursuing)
        final pursuing = army1.destination == army2.origin || army2.destination == army1.origin;
        final closeEnough = (army1.turnsUntilArrival - army2.turnsUntilArrival).abs() <= 1;

        final shouldIntercept = oppositeDirection ||
            (sameDestination && bothCloseToArrival) ||
            (pursuing && closeEnough && army1.turnsUntilArrival <= 1);

        if (shouldIntercept) {
          game.addTurnEvent(ArmyInterceptedEvent(
            army1Name: army1.name,
            army2Name: army2.name,
          ));
          _resolveFieldBattle(army1.id, army2.id);
          interceptedArmyIds.add(army1.id);
          interceptedArmyIds.add(army2.id);
        }
      }
    }
  }

  void _resolveFieldBattle(String army1Id, String army2Id) {
    final game = GameManager.shared;

    final army1 = game.armies.cast<Army?>().firstWhere((a) => a!.id == army1Id, orElse: () => null);
    final army2 = game.armies.cast<Army?>().firstWhere((a) => a!.id == army2Id, orElse: () => null);

    if (army1 == null || army2 == null) return;

    _combatEngine.resolveCombat(
      attackers: army1.units,
      defenders: army2.units,
      map: game.map,
      defendingVillage: null,
    );

    army1.removeDeadUnits();
    army2.removeDeadUnits();

    if (army1.units.isEmpty) {
      game.removeArmy(army1.id);
    } else {
      game.updateArmy(army1);
    }

    if (army2.units.isEmpty) {
      game.removeArmy(army2.id);
    } else {
      game.updateArmy(army2);
    }
  }

  void _processArmyMovement() {
    final game = GameManager.shared;
    final arrivedArmies = <(Army, Village)>[];

    // Advance all marching armies
    for (var i = 0; i < game.armies.length; i++) {
      if (game.armies[i].isMarching) {
        game.armies[i].advanceMarch();

        if (!game.armies[i].isMarching) {
          final destId = game.armies[i].stationedAt;
          if (destId != null) {
            final destination = game.map.villages.cast<Village?>().firstWhere(
                  (v) => v!.id == destId,
                  orElse: () => null,
                );
            if (destination != null) {
              arrivedArmies.add((game.armies[i], destination));
            }
          }
        }
      }
    }

    // Process arrivals
    for (final (army, destination) in arrivedArmies) {
      if (army.owner != destination.owner) {
        _resolveCombat(army, destination);
      } else {
        game.mergeArmiesAt(destination.id, army.owner);
        game.addTurnEvent(ArmyArrivedEvent(armyName: army.name, destination: destination.name));
      }
    }
  }

  void _resolveCombat(Army attacker, Village village) {
    final game = GameManager.shared;

    // Mark village as under siege
    village.underSiege = true;

    // Get defending armies
    final defenderArmies = game.getArmiesAt(village.id).where((a) => a.owner == village.owner).toList();
    final defenderUnits = defenderArmies.expand((a) => a.units).toList();

    final result = _combatEngine.resolveCombat(
      attackers: attacker.units,
      defenders: defenderUnits,
      map: game.map,
      defendingVillage: village,
    );

    // Update attacker army
    attacker.removeDeadUnits();
    if (attacker.units.isEmpty) {
      game.removeArmy(attacker.id);
    } else {
      game.updateArmy(attacker);
    }

    // Update defender armies
    for (final defArmy in defenderArmies) {
      game.removeArmy(defArmy.id);
    }
    final survivingDefenders = defenderUnits.where((u) => u.isAlive).toList();
    if (survivingDefenders.isNotEmpty) {
      game.createArmy(survivingDefenders, village.id, village.owner);
    }

    // Check for conquest
    final canConquer = result.attackerWon && survivingDefenders.isEmpty;

    if (canConquer) {
      final oldOwner = village.owner;
      village.owner = attacker.owner;
      village.population = (village.population * 0.8).toInt();
      village.happiness = max(30, village.happiness - 20);
      village.garrisonStrength = 5;
      village.underSiege = false;
      game.updateVillage(village);

      // Station attacking army
      final attackingArmy = game.armies.cast<Army?>().firstWhere((a) => a!.id == attacker.id, orElse: () => null);
      if (attackingArmy != null) {
        attackingArmy.station(village.id);
        game.updateArmy(attackingArmy);
      }

      if (attacker.owner == 'player') {
        game.addTurnEvent(VillageConqueredEvent(villageName: village.name));
      } else if (oldOwner == 'player') {
        game.addTurnEvent(VillageLostEvent(villageName: village.name));
      }
    } else if (result.attackerWon) {
      // Won battle but defenders survived
      final garrisonDamage = max(3, village.garrisonStrength ~/ 2);
      village.damageGarrison(garrisonDamage);
      game.updateVillage(village);

      if (attacker.owner == 'player') {
        game.addTurnEvent(BattleWonEvent(location: village.name, casualties: result.attackerCasualties));
      } else if (village.owner == 'player') {
        game.addTurnEvent(BattleLostEvent(location: village.name, casualties: result.defenderCasualties));
      }
    } else {
      // Lost the battle
      village.underSiege = false;
      game.updateVillage(village);

      if (attacker.owner == 'player') {
        game.addTurnEvent(BattleLostEvent(location: village.name, casualties: result.attackerCasualties));
      } else if (village.owner == 'player') {
        game.addTurnEvent(BattleWonEvent(location: village.name, casualties: result.defenderCasualties));
      }
    }
  }

  void _processAITurns() {
    final game = GameManager.shared;
    final aiPlayers = game.players.where((p) => !p.isHuman && !p.isEliminated).toList();

    for (final aiPlayer in aiPlayers) {
      _aiEngine.executeAITurn(aiPlayer, game.map);
    }
  }

  void _detectIncomingEnemies() {
    final game = GameManager.shared;
    final playerVillages = game.getPlayerVillages('player');

    for (final army in game.armies) {
      if (army.owner == 'player') continue;
      if (!army.isMarching) continue;

      final destId = army.destination;
      if (destId == null) continue;

      if (playerVillages.any((v) => v.id == destId)) {
        final destVillage = game.map.villages.cast<Village?>().firstWhere(
              (v) => v!.id == destId,
              orElse: () => null,
            );
        if (destVillage != null) {
          game.addTurnEvent(EnemyApproachingEvent(
            enemyName: army.name,
            target: destVillage.name,
            turns: army.turnsUntilArrival,
          ));
        }
      }
    }
  }

  void _checkVictory() {
    final game = GameManager.shared;

    for (var i = 0; i < game.players.length; i++) {
      final playerVillages = game.map.villages.where((v) => v.owner == game.players[i].id);
      if (playerVillages.isEmpty && !game.players[i].isEliminated) {
        game.players[i] = game.players[i].copyWith(isEliminated: true);
      }
    }
  }
}
